### PR TITLE
30 update codebase to work with Python 3.8 (and thus ase v3.23)

### PR DESCRIPTION
**Pull request by logsdail**
Testing to updated codebase with ASE release

### CODE REVIEW

**Comment by logsdail** [comment]
Noting here that the docs runs with Python 3.9 - should we standardise to this version, perhaps?

**Comment by logsdail** [comment]
So I think after discussion today at a Group meeting, this has become an "update Python version to the most suitable for our HPC architectures". 

Current Python compatibility of code: Python 3.7 (i.e., CI tests work to this version)

Current HPC system Python provisions:
Hawk: 3.7 (used currently, default), 3.9.2, 3.10.4, 3.11.2
Isambard: 3.7.3.2 (used currently), 3.8.5.1 (default)
Young: 3.6 (used currently), 3.7.0, 3.7.2, 3.7.4, 3.8.0, 3.8.6 (default), 3.9.0, 3.9.1, 3.9.6, 3.9.10, 3.11.3, 3.11.4
ARCHER2: 3.6.15 (used currently) 3.9.13.1 (default) 3.10.10

**Comment by logsdail** [comment]
> So I think after discussion today at a Group meeting, this has become an "update Python version to the most suitable for our HPC architectures".
> 
> Current Python compatibility of code: Python 3.7 (i.e., CI tests work to this version)
> 
> Current HPC system Python provisions: Hawk: 3.7 (used currently, default), 3.9.2, 3.10.4, 3.11.2 Isambard: 3.7.3.2 (used currently), 3.8.5.1 (default) Young: 3.6 (used currently), 3.7.0, 3.7.2, 3.7.4, 3.8.0, 3.8.6 (default), 3.9.0, 3.9.1, 3.9.6, 3.9.10, 3.11.3, 3.11.4 ARCHER2: 3.6.15 (used currently) 3.9.13.1 (default) 3.10.10

My take home from this is the minimum we should go to is 3.8 or 3.9, but certainly not towards the high numbers.

**Comment by logsdail** [comment]
@PavelStishenko as suggested I've put a tag on the mainline version, and once merged we'll bumpy the minor version number.

File: examples/analyse_forces.py
Comment (ikowalec) [inline examples/analyse_forces.py:37]: `atoms.get_forces()` will calculate the forces, previously, the `atoms.calc.forces` retrieved an empty list from the results dictionary of the calculator object

Comment (ikowalec) [reply examples/analyse_forces.py:37]: I suggest removing the print statement completely

File: .github/workflows/main.yml
Comment (ikowalec) [inline .github/workflows/main.yml:56]: Do we want to keep the debug?

File: examples/analyse_counterpoise.py
Comment (ikowalec) [inline examples/analyse_counterpoise.py:4]: The description is required, but technically an issue outside of the goal of this pull request.

File: examples/build_symmetry_mirror.py
Comment (ikowalec) [inline examples/build_symmetry_mirror.py:3]: not sure, about this change since utility functions reside in the `utils` folder

File: carmm/utils/python_env_check.py
Comment (ikowalec) [inline carmm/utils/python_env_check.py:6]: Can the descriptions be moved to the respective functions?

File: carmm/run/workflows/react.py
Comment (ikowalec) [inline carmm/run/workflows/react.py:291]: Presumably leftover from the debugging process

**Review by ikowalec — APPROVED** [review]
There is some tidy up to do, but that is not related to the compatibility issues, which have been resolved successfully.

File: examples/analyse_counterpoise.py
Comment (logsdail) [reply examples/analyse_counterpoise.py:34]: No - removing

File: examples/analyse_forces.py
Comment (logsdail) [reply examples/analyse_forces.py:37]: Fair - and solution agreed.

File: .github/workflows/main.yml
Comment (logsdail) [reply .github/workflows/main.yml:56]: I think no harm to keep it - means we can see versions if any questions arise, and low cost.

File: carmm/run/workflows/react.py
Comment (logsdail) [reply carmm/run/workflows/react.py:291]: Removed.

File: carmm/utils/python_env_check.py
Comment (logsdail) [reply carmm/utils/python_env_check.py:28]: Resolved with new commit

**Comment by logsdail** [comment]
@ikowalec thanks for comments - all addressed. Does someone else want to do a second review, please?

**Comment by logsdail** [comment]
@GaryLZW I simplified the `analyse_counterpoise` test, as the checking for ASE now happens in `get_aims_calculator`. The CI tests pass but you might want to check if the code works correctly for further use cases.

File: carmm/run/aims_calculator.py
Comment (GaryLZW) [inline carmm/run/aims_calculator.py:55]: Hi Andy. I am not sure about this line. My understanding is this.
If ase version is 3.22, `AimsProfile(command='xc')` is an Aims calculator. Will this pollute the `control.in` file `fhi_calc` generates? I didn't find where they remove the non-existing parameters in the old write_control function.
If ase version is 3.23, `AimsProfile(command='xc')` provides a profile for `fhi_calc`. This seems to let dry run work. But does it prevent the `fhi_calc` to load commands and paths when doing real calculations?

**Comment by GaryLZW** [comment]
> @GaryLZW I simplified the `analyse_counterpoise` test, as the checking for ASE now happens in `get_aims_calculator`. The CI tests pass but you might want to check if the code works correctly for further use cases.

This looks fine to me^^

File: carmm/run/aims_calculator.py
Comment (logsdail) [reply carmm/run/aims_calculator.py:55]: Fair questions, but hard to answer as you are basically asking for situations outside the boundaries of our CI tests. The only way we can find out is by testing, and then improving the CI tests if things don't work.

Comment (logsdail) [reply carmm/run/aims_calculator.py:55]: For actual use with ASE v3.23, all we have currently is the CI tests - we'd have to deploy and use to see if it works, or doesn't, and then resolve as necessry.

Comment (logsdail) [reply carmm/run/aims_calculator.py:55]: (We can't fix bugs we don't know about!)

Comment (logsdail) [reply carmm/run/aims_calculator.py]: @GaryLZW is this at least clearer now in implementation? And we can work to see how things work when actually deployed?

Comment (GaryLZW) [reply carmm/run/aims_calculator.py]: Hi Andy. Yes. The code looks clearer to me now. Happy to resolve further issues after deploying.

**Review by GaryLZW — APPROVED** [review]

File: carmm/run/aims_calculator.py
Comment (GabrielBram) [reply carmm/run/aims_calculator.py]: I think a more permanent solution should be feasible when we get around to re-working ReactAims and related calculator handling routines for ASE 3.23 - I can imagine we can create a dummy profile that is loaded for CI testing, and loads the user-generated profile otherwise.

Comment (logsdail) [reply carmm/run/aims_calculator.py]: Agreed - there is a better solution out there. I think we also need to make the executable commands internal, rather than system environment settings.

**Comment by FirasAssa** [comment]
During my project the following changes were made:
The err_handler folder was removed and a utils folder was created in its place
A python_env_check file was created with python_env_check() and is_env_python() inside
Implemented python_env_check() in anylyse_forces example 
Implemented python_env_check() in analyse_calculator
Restricted mace-torch to 0.3.4, because it was failing on v3.7 
Updated CatLearn to the most recent version to fix import error in mlneb.py, by getting the version of CatLearn directly from Git 

Thinks still to do after the project:
In python v3.8 ci-test there are 4 places where .BadConfiguration Error occurs. Presumably it is the same type of problem that can be fixed in one go
Errors:
ERROR analyse_counterpoise.py - ase.calculators.calculator.BadConfiguration: No configuration of aims
ERROR run_aims.py - ase.calculators.calculator.BadConfiguration: No configuration of aims
ERROR run_workflows_ReactAims.py - ase.calculators.calculator.BadConfiguration: No configuration of aims
ERROR run_workflows_ReactAims_parallel.py - ase.calculators.calculator.BadConfiguration: No configuration of aims

**Comment by logsdail** [comment]
[like]  Andrew Logsdail reacted to your message:
________________________________

**Pull request merged**
